lisa/tools/fio.py

```diff
@@ -107,20 +111,87 @@ def launch(
             cwd=cwd,
             timeout=ssh_timeout,
         )
-        matched_results = self._result_pattern.match(result.stdout)
+        fio_result = self.get_result_from_raw_output(
+            mode, result.stdout, iodepth, numjob
+        )
+        return fio_result
+
+    def launch_async(
+        self,
+        name: str,
+        filename: str,
+        mode: str,
+        iodepth: int,
+        numjob: int,
+        time: int = 120,
+        block_size: str = "4K",
+        size_gb: int = 0,
+        direct: bool = True,
+        gtod_reduce: bool = False,
+        ioengine: str = "libaio",
+        group_reporting: bool = True,
+        overwrite: bool = False,
+        time_based: bool = False,
+        cwd: Optional[pathlib.PurePath] = None,
+    ) -> Process:
+        cmd = self._get_command(
+            name,
+            filename,
+            mode,
+            iodepth,
+            numjob,
+            time,
+            block_size,
+            size_gb,
+            direct,
+            gtod_reduce,
+            ioengine,
+            group_reporting,
+            overwrite,
+            time_based,
+        )
+        process = self.run_async(
+            cmd,
+            force_run=True,
+            sudo=True,
+            cwd=cwd,
+        )
+        return process
+
+    def wait_start(self, process: Process, timeout: int = 300) -> bool:
+        # FIO output emits lines of the form when it is running
+        # (f=10): [M(10)][24.9%][r=70.0MiB/s,w=75.0MiB/s][r=70,w=75 IOPS][eta 03m:46s]
+        # check if stdout buffers contain the string "eta" to determine if it is running
+        start_time = time.time()
+        while time.time() - start_time < timeout:
+            if "eta" in process._stdout_writer._buffer:
```

Comment:
The private method shouldn't be accessed outside. Please add a method called `stdout_contains` in `Process`.

Reply:
I can put this method in `Process` but we would still need to access the `_buffer` object which is internal to the spur library. Another issue here is that `word in string` test is not very general and might not be applicable in other scenarios.

Reply:
The buffer may be huge, and I'm not sure if it's format is bytes or str. So I don't want it to be copied for each comparison, contains is pretty lightweight. Let's see if there are more requirements in future. We can update it to support regexp if there are.

Reply:
I found a bug where the buffer was getting cleared faster than we were reading it. 0.1 second interval worked well but this problem came up when we moved to 1 sec interval. I have modified LogWriter to include a flush timeout and then read buffer faster than this timeout

Reply:
That's my concern, the buffer may be updated very fast, it's not a reliable way. It may need to add a new log handler on each process to save stdout to a string, and do search.

Reply:
Let's discuss it offline to find the best way forward.